src/main/java/org/apache/mesos/hdfs/Scheduler.java

```diff
@@ -73,18 +81,178 @@ public void registered(SchedulerDriver driver, FrameworkID frameworkId, MasterIn
     try {
       persistentState.setFrameworkId(frameworkId);
     } catch (InterruptedException | ExecutionException e) {
+      log.error("Error setting framework id in persistent state", e);
       throw new RuntimeException(e);
     }
     log.info("Registered framework frameworkId=" + frameworkId.getValue());
+    //reconcile tasks upon registration
+    reconcileTasks(driver);
   }
 
   @Override
   public void reregistered(SchedulerDriver driver, MasterInfo masterInfo) {
-    log.info("Reregistered framework.");
+    log.info("Reregistered framework: starting task reconciliation");
+    // reconcile tasks upon reregistration
+    reconcileTasks(driver);
+  }
+
+  @Override
+  public void statusUpdate(SchedulerDriver driver, TaskStatus status) {
+    log.info(String.format(
+        "Received status update for taskId=%s state=%s message='%s' stagingTasks.size=%d",
+        status.getTaskId().getValue(),
+        status.getState().toString(),
+        status.getMessage(),
+        liveState.getStagingTasksSize()));
+
+    if (!isStagingState(status)) {
+      liveState.removeStagingTask(status.getTaskId());
+    }
+
+    if (isTerminalState(status)) {
+      liveState.removeRunningTask(status.getTaskId());
+      persistentState.removeTaskId(status.getTaskId().getValue());
+      if (reconciliationComplete()) {
```

Comment:
I think the logic is correct; I was just saying that it was non-obvious and deserves a comment.
You want to correct the phase if ((RECONCILING && reconciliationComplete) || OTHER_STATE), where reconciliationComplete is always true if you're already in any state other than RECONCILING. Thus, you should call correctCurrentPhase for any terminalStatus if (reconciliationComplete).